crypto/evp/evp_locl.h

```diff
@@ -65,6 +65,64 @@ struct evp_Encode_Ctx_st {
 typedef struct evp_pbe_st EVP_PBE_CTL;
 DEFINE_STACK_OF(EVP_PBE_CTL)
 
-int is_partially_overlapping(const void *ptr1, const void *ptr2, int len);
-int is_any_overlapping(const void *ptr1, int len1, const void *ptr2, int len2);
-int is_pointer_offset_by_bias(const void *ptr1, const void *ptr2, int bias);
+/*
+ * According to the letter of standard difference between pointers
+ * is specified to be valid only within same object. This makes
+ * it formally challenging to determine if input and output buffers
+ * are not partially overlapping with standard pointer arithmetic.
+ */
+#ifdef PTRDIFF_T
+# undef PTRDIFF_T
+#endif
+#if defined(OPENSSL_SYS_VMS) && __INITIAL_POINTER_SIZE==64
+/*
+ * Then we have VMS that distinguishes itself by adhering to
+ * sizeof(size_t)==4 even in 64-bit builds, which means that
+ * difference between two pointers might be truncated to 32 bits.
+ * In the context one can even wonder how comparison for
+ * equality is implemented. To be on the safe side we adhere to
+ * PTRDIFF_T even for comparison for equality.
+ */
+# define PTRDIFF_T uint64_t
+#else
+# define PTRDIFF_T size_t
+#endif
+
+static ossl_inline
+int is_partially_overlapping(const void *ptr1, const void *ptr2, int len)
+{
+    PTRDIFF_T diff = (PTRDIFF_T)ptr1 - (PTRDIFF_T)ptr2;
+    /*
+     * Check for partially overlapping buffers. [Binary logical
+     * operations are used instead of boolean to minimize number
+     * of conditional branches.]
+     */
+    int overlapped = (len > 0) & (diff != 0) & ((diff < (PTRDIFF_T)len) |
+                                                (diff > (0 - (PTRDIFF_T)len)));
+
+    return overlapped;
+}
+
+static ossl_inline
+int is_any_overlapping(const void *ptr1, int len1, const void *ptr2, int len2)
+{
+    PTRDIFF_T diff = (PTRDIFF_T)ptr1 - (PTRDIFF_T)ptr2;
+    /*
+     * Same as above, but check for any overlapping between
+     * [ptr1..ptr1+len1[ and [ptr2..ptr2+len2[
+     */
+    int overlapped = (len1 > 0) & (len2 > 0)
+                     & ((diff < (PTRDIFF_T)len2)
+                        | (diff > (0 - (PTRDIFF_T)len1)));
+
+    return overlapped;
+}
+
+static ossl_inline
+int is_pointer_offset_by_bias(const void *ptr1, const void *ptr2, int bias)
+{
+    PTRDIFF_T diff = (PTRDIFF_T)ptr1 - (PTRDIFF_T)ptr2;
+
+    diff -= bias;
+    return diff == 0;
```

Comment:
On a **side** note I kind of liked original better, i.e. straight comparison of one pointer plus offset with another pointer. It was more readable, as here you have to work unnecessary hard to confirm that it's correct. Once again, side[!] note :-)